src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs

```diff
@@ -24,12 +29,86 @@ public static HashProvider CreateMacProvider(string hashAlgorithmId, ReadOnlySpa
+                    hashAlgorithmId,
+                    BCryptOpenAlgorithmProviderFlags.None);
+
+                int hashSize;
+
+                NTSTATUS ntStatus = Interop.BCrypt.BCryptGetProperty(
+                    cachedAlgorithmHandle,
+                    Interop.BCrypt.BCryptPropertyStrings.BCRYPT_HASH_LENGTH,
+                    &hashSize,
+                    sizeof(int),
+                    out _,
+                    0);
+
+                if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+
+                if (destination.Length < hashSize)
+                    throw new CryptographicException();
+
+                if (!s_UseCompatOneShot)
                 {
-                    hashProvider.AppendHashData(source);
-                    return hashProvider.FinalizeHashAndReset(destination);
+                    try
+                    {
+                        fixed (byte* pSource = source)
+                        fixed (byte* pDestination = destination)
+                        {
+                            ntStatus = Interop.BCrypt.BCryptHash(cachedAlgorithmHandle, null, 0, pSource, source.Length, pDestination, hashSize);
+
+                            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                            {
+                                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                            }
+                        }
+                    }
+                    catch (EntryPointNotFoundException)
+                    {
+                        s_UseCompatOneShot = true;
+                        HashUpdateAndFinish(cachedAlgorithmHandle, hashSize, source, destination);
+                    }
+                }
+                else
+                {
+                    HashUpdateAndFinish(cachedAlgorithmHandle, hashSize, source, destination);
+                }
+
+                return hashSize;
+            }
+
+            private static void HashUpdateAndFinish(
+                SafeBCryptAlgorithmHandle algHandle,
+                int hashSize,
+                ReadOnlySpan<byte> source,
+                Span<byte> destination)
+            {
+                NTSTATUS ntStatus = Interop.BCrypt.BCryptCreateHash(
+                    algHandle,
+                    out SafeBCryptHashHandle hHash,
+                    IntPtr.Zero,
+                    0,
+                    default,
+                    0,
+                    BCryptCreateHashFlags.None);
+
+                if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+
+                using (hHash)
+                {
+                    ntStatus = Interop.BCrypt.BCryptHashData(hHash, source, source.Length, 0);
+
+                    if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                        throw Interop.BCrypt.CreateCryptographicException(ntStatus);
```

Comment:
Braces on all the blocks in this change (since none are preamble argument validation) would make me happier.